data/firestore/src/main/java/io/github/droidkaigi/confsched2020/data/firestore/internal/FirestoreImpl.kt

```diff
@@ -83,6 +120,40 @@ internal class FirestoreImpl @Inject constructor() : Firestore {
         firebaseAuth.signInAnonymously().await()
         Timber.debug { "signInIfNeeded end" }
     }
+
+    private fun getThumbsUpCounterRef(sessionId: SessionId): CollectionReference {
+        return FirebaseFirestore
+            .getInstance()
+            .collection("confsched/2020/sessions/${sessionId.id}/thumbsup_counters")
+    }
+
+    private suspend fun createShardsIfNeeded(counterRef: CollectionReference) {
+        if (!counterRef.fastGet().isEmpty) {
+            Timber.debug { "createShardsIfNeeded shards already exist" }
+            return
+        }
+
+        val tasks = arrayListOf<Task<Void>>()
+        (0 until NUM_SHARDS).forEach {
+            val makeShard = counterRef
+                .document(it.toString())
+                .set(mapOf(SHARDS_COUNT_KEY to 0))
+            tasks.add(makeShard)
+        }
+
+        try {
+            Tasks.whenAll(tasks).await()
+        } catch(e: Exception) {
+            println("⭐" + e.message)
+        }
+        Timber.debug { "createShardsIfNeeded creating shards completed" }
+    }
+
+    companion object {
+        const val NUM_SHARDS = 10
```

Comment:
If there seems to be no problem in speed, I want to change it to 5 for cost.

```suggestion
        const val NUM_SHARDS = 5
```

Reply:
5 shards are able to update 5 times for 1 second.
That means that throughput is 5.

1. 100 persons tap thumbs-up button 10 times for 60 seconds. (very popular session?)
Requirement throughput
100 * 10 / 60 ≒ 16.7

1. 100 persons tap thumbs-up button 5 times for 60 seconds. (busy?)
Requirement throughput
100 * 5 / 60 ≒ 8.33

1. 50 persons tap thumbs-up button 5 times for 60 seconds. (normaly?)
Requirement throughput
50 * 5 / 60 ≒ 4.16

Umm... 
I think enough speed, but I do not have confidence😅

Reply:
If firestore is free, I did so, but I am concerned that the price will be expensive because every time the value flows, the number of reads runs to everyone.

Reply:
I forgot firebase pricing.
I have to reduce accesses of firestore.
I'm thinking it a little more.

Reply:
Please tell me your opinion 🙏 I think we can thumbs up 300 / minutes. Is it not enough?

Reply:
Let's calculate a little.
100 people are watching * 10 shards * (100 people * 100 times) * 40 sessions = 400 million read
Read Document $ 0.038 per 100,000 Documents
0.038 * (100 * 10 * 100 * 100 * 40) / 100000
= $ 152.0

(100 people * 100 times) * 40 sessions = 400000
Write Document $ 0.115 per 100,000 Documents

0.115 * 100*100*40 / 100000
= $ 0.46

About 15,000 yen (  Organizer @mhidaka says it is OK)

But you may be able to do something. 👀 
For example, if you have 9 shards and monitor only 3 shards, you can multiply by 3 to get 1/3.

Reply:
Even if this does not reflect your thumbs-up, it may not be good.

Reply:
I have ideas for reduce access to firestore.
So I think 5 shards is good.
https://github.com/DroidKaigi/conference-app-2020/issues/429#issuecomment-582482685

I apply your change!